Add wrapper structure to work on (former #66)

rrtoledo opened the pull request:
## Content

This PR aims at restructuring the Centralized Telescope code to have a user facing structure from which we can directly prove and verify.. More particularly,
- Moves code in implementation block
- Moves implementation block in struct file
- Renamed the user facing wrapper structure
- Removed the user parameter structure and renamed the internal parameter structure

Additionnally, this PR
- reintroduced the cases for the parameter generation
- adding more comprehensive documentation for public functions
- Move Hash type to utils/types.rs
- Fixed clippy error from tests
- Set Proof.prove/verify to public for consistency with Params

## Pre-submit checklist

- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
- PR
    - [x] No clippy warnings in the CI
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- Documentation
    - [x] Update README file (if relevant)
    - [x] Update documentation website (if relevant)

## Comments

Open question:
- Should we add getters to access struct parameters or keep the struct fields public?

## Issue(s)

<!-- The issue(s) this PR relates to or closes -->

Relates to #59

tolikzinovyev commented:
This PR is large and I still don't agree with some of the changes (e.g., structs with public fields and impl). Suggest splitting it into smaller PRs. For example, I can quickly approve public functions docs. Refactoring of parameter derivation cases - I'm currently unable to see what it improves exactly, could you explain? Ideally, such explanation should be in the PR description. There might be other changes I'm missing, it is very hard see...

rrtoledo commented:
> This PR is large and I still don't agree with some of the changes (e.g., structs with public fields and impl). Suggest splitting it into smaller PRs. For example, I can quickly approve public functions docs. Refactoring of parameter derivation cases - I'm currently unable to see what it improves exactly, could you explain? Ideally, such explanation should be in the PR description. There might be other changes I'm missing, it is very hard see...

Sure I have just added some comments on the PR and split the Centralized lottery and simple lottery parts in 2 (c.f. PR #118).

The changes look big but except for the reintroduction of cases there is no really big modifications (look at the commits' titles you'll see!), most of the changes are due to moving code around.
The parameters cases simplifies the param code, gives more structure to the code, and is needed for #117.

tolikzinovyev commented:
> > This PR is large and I still don't agree with some of the changes (e.g., structs with public fields and impl). Suggest splitting it into smaller PRs. For example, I can quickly approve public functions docs. Refactoring of parameter derivation cases - I'm currently unable to see what it improves exactly, could you explain? Ideally, such explanation should be in the PR description. There might be other changes I'm missing, it is very hard see...
> 
> @tolikzinovyev Sure I have just added some comments on the PR and split the Centralized lottery and simple lottery parts in 2 (c.f. PR #118).
> 
> The changes look big but except for the reintroduction of cases there is no really big modifications (look at the commits' titles you'll see!), most of the changes are due to moving code around.
> 
>     * The implementation blocks were agreed by both @curiecrypt and @djetchev several weeks ago in a team meeting you attended.
> 
>     * I am more than happy to add getters functions and set the fields to private
> 
>     * The parameters cases simplifies the param code, gives more structure to the code, and is needed for [Adding setup_safe functions #117](https://github.com/cardano-scaling/alba/pull/117).

Thanks for the PR description! It has nine bullet points and most of them can be a separate PR. Right now, this is still very hard to review.

I don't remember a definite decision about impl blocks, but I would be happy to see what structs with private fields would look like. Please make a PR with only that change or alternatively write an overview of the change somewhere.

I'm still not convinced we need #117.

I would suggest to first make PRs with public functions docs and one-two-line changes that you made, as those would be easy to review and merge. Thanks!
